### ISSUE TITLE
PVF: Compromised artifact file integrity can lead to disputes

### ISSUE THREAD

**Issue by mrcnski** [paritytech/polkadot-sdk#677]
# ISSUE

## Overview

If we try to execute a corrupt PVF artifact it will likely lead to a dispute. See the comment below for an example of this. 

A simple mitigation is to just put the hash of the file contents in the filename. Artifacts are not that big, on the order of several dozen MB (I think?). If we use something like https://github.com/BLAKE3-team/BLAKE3, which claims 6866 MiB/s throughput, then the time spent hashing would barely register. 

## Original comment

As far as I understand, the main concern about not clearing the artifact cache is:
1) Unsafe functions are used to import the artifact
2) [Their safety guarantees](https://github.com/paritytech/polkadot/blob/2bbef460d24f44aeba9d4014062bb8c631327496/node/core/pvf/src/executor_intf.rs#L199) rely on the fact the node has produces them with `prepare()` and they are not just some random files put there by someone.

Not that we can guarantee nobody will overwrite them during the node run time. I believe the cache pruning was implemented as a measure to mitigate those unwanted outcomes. But I don't have a strong opinion if it makes any sense. If someone wants to screw things up, he will.

_Originally posted by @s0me0ne-unkn0wn in https://github.com/paritytech/polkadot-sdk/issues/685

**Comment by s0me0ne-unkn0wn** [comment]
Not sure I follow. You mean, hash a prepared artifact right after the preparation, include its hash into its name, and check the hash before execution? But how do you know the name of the artifact you're going to open for execution, then? Sounds like a chicken and egg situation.

**Comment by mrcnski** [comment]
> Not sure I follow. You mean, hash a prepared artifact right after the preparation, include its hash into its name, and check the hash before execution? But how do you know the name of the artifact you're going to open for execution, then? Sounds like a chicken and egg situation.

I was thinking on node startup, we just list all the artifacts in the cache path, check consistency with the hash, and load them into the artifacts table? As mentioned above, hashing before each execution would be too much of a performance hit.

**Comment by s0me0ne-unkn0wn** [comment]
Aha, thanks for the elaboration, that does make sense. But that would protect from accidental artifact corruption but not from malicious action. On the other hand, I don't see any easy way to protect against malicious artifacts while still keeping them on disk anyway. (Sign them with the validator key maybe? But that would require some effort)

**Comment by eagr** [comment]
Please help me understand. What would happen if someone deliberately swaps the cache after preparation? And what if they do that in a large scale? Is there any incentive of doing that?

**Comment by burdges** [comment]
> My main concern is avoiding unneeded disputes due to a hardware failure (corrupted disk)

These cause invalid disputes, which disables the validator.  We need not punish those too heavily, unless correlated across validators, so maybe the solution is just to take this into consideration in the slashing level?

**Comment by mrcnski** [comment]
@burdges Related to that, another idea we had with @eskimor is to only hash the file if a candidate is found to be invalid, after the fact. Reason being that this is a rare case so this additional check shouldn't affect performance too much. If an artifact was corrupted it is very unlikely to be found valid, so we shouldn't need to check file integrity in that case.

**Comment by s0me0ne-unkn0wn** [comment]
@mrcnski sounds like a really good idea!

**Comment by mrcnski** [comment]
Yes, I filed a follow-up here: https://github.com/paritytech/polkadot-sdk/issues/2399#issuecomment-1826787612 Please coordinate with @Jpserrat as he commented there, I'm not sure if he started on this yet or not!

**Comment by burdges** [comment]
Yes blake3 is fast, but it's speed beyond blake2 comes from merkle trees being more parallelizable, including both SIMD and multiple threads.  The SIMD is a clear win, but polkadot should be using those other CPU cores, although blake3 improves cache locallity maybe.

Anyways, we might still prefer to do this check only when some validation fails.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/2025-05-03-polkadot-parachain-block-time-degradation/12963/1

**Comment by rphmeier** [comment]
Checksumming seems like a really good idea!

Artifacts are pretty small (a few MB), and should be very quick to hash. The overhead in loading them from disk will probably be the I/O work. And checking the checksum can probably be moved out of the critical path if really needed.

I think a cryptographic hash like blake3 is probably overkill here and using a hash like twox is somewhat better because it has even higher throughput. https://github.com/shepmaster/twox-hash

Before downloading the PVF in the first place we rely on the runtime to check the code hash versus a blake3 hash. So nodes already have checked the validity with a cryptographic checksum at this point in the 'PVF supply chain'

Though if your threat model does involve attackers surreptitiously editing the files on disk with malicious code that collides with the previous hash, you can have each node parameterize their twox instance with a custom seed (so every node will generate different hashes and attackers would have to find collisions per-machine). Probably overkill but worth putting out there.

**Comment by burdges** [comment]
Yes, there are some attack vectors like someone prepares a PVF in advance which then a fresh validator launches on, which do benefit from the hash being specific to some secret from the TPM or something.  I doubt blake3 being slower than twox matters if this only occurs when nodes start up, but yes you can use something faster like SipHash in that setting.  

I'm unsure xxhash (twox) works though.  It's maybe fine if the file names remain secret.  If the artifact file names leak via logs, discussions, etc. then the twox key could maybe be extracted, and they could create validator specific collisions.  If the artifact name still contains the blake2 hash of the source wasm/pvm too, then the adversary cannot get this onto the validator by a simply on-chain upgrade, and must somewhat hack the validator anyways.
